libsolutil/DisjointSet.cpp

```diff
@@ -0,0 +1,110 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/DisjointSet.h>
+
+#include <liblangutil/Exceptions.h>
+
+#include <numeric>
+
+using namespace solidity::util;
+
+ContiguousDisjointSet::ContiguousDisjointSet(size_t const _numNodes):
+	m_parents(_numNodes),
+	m_neighbors(_numNodes),
+	m_sizes(_numNodes, static_cast<value_type>(1)),
+	m_numSets(_numNodes)
+{
+	// each is their own neighbor and parent
+	std::iota(m_parents.begin(), m_parents.end(), 0);
+	std::iota(m_neighbors.begin(), m_neighbors.end(), 0);
+}
+
+size_t ContiguousDisjointSet::numSets() const { return m_numSets; }
+
+ContiguousDisjointSet::value_type ContiguousDisjointSet::find(value_type const _element) const
+{
+	solAssert(_element < m_parents.size());
+	// path halving
+	value_type rootElement = _element;
+	while (rootElement != m_parents[rootElement])
+	{
+		m_parents[rootElement] = m_parents[m_parents[rootElement]];
+		rootElement = m_parents[rootElement];
+	}
+	return rootElement;
+}
+
+bool ContiguousDisjointSet::merge(value_type const _x, value_type const _y, bool const _mergeBySize)
+{
+	auto xRoot = find(_x);
+	auto yRoot = find(_y);
+
+	if (xRoot == yRoot)
+		return false;  // we're done, nothing to merge here
```

Comment:
I believe this is done because the root nodes in this data structure represent the sets they contain, so two nodes are in the same set if and only if the roots of the trees containing the nodes are equal.

Reply:
It is not strictly required and indeed, for now I just used it for testing. At first I was thinking it could give you some performance benefit to return the status code here but in reality this isn't even true - let's remove it:)

Reply:
Yup, I know that - I'm just arguing that the `merge` function (which should be called `union`) doesn't need to return anything, i.e. it should be `void`. I'm simply asking whether having it returning a boolean indicating that a union operation was performed (or whether the two elements were already in the same set) is going to be relevant information when used within it the liveness analysis?

Reply:
I would have called it union if union wasn't a reserved keyword in c++ ;)

Reply:
`uNiOn`. I personally prefer merge, because it's more comp sci and less math :D

Reply:
boo! :D i kind of like your suggestion, has that sarcastic aftertaste.. let's make a `uNi0n` of these two sets!